File: packages/service/src/domain/commands/model/command-processor.ts
/* eslint-disable max-classes-per-file */
import { ExecutionContext } from '@service/domain/commands/execution-context';
import { Command } from '@service/domain/commands/model/command';
import { CommandProcessingResponse } from '@service/domain/commands/model/command-processing-response';

export abstract class CommandProcessor {
  abstract execute(command: Command, context: ExecutionContext): Promise<CommandProcessingResponse>

  abstract get key(): string;
  abstract get shortKey(): (string | null);
  abstract get helpMessage(): string;
  abstract get exampleUsages(): string[];
  abstract get parameters(): CommandParameters;
}

export interface CommandParameters {
  parameters: CommandParameter[],
  delimeter: (string | null),
}

export interface CommandParameter {
  required: boolean,
  names: string[],
}

export class CommandParametersBuilder {
  parameters: CommandParameter[] = [];
  delimeter: (string|null) = null;

  withRequired(name: string): CommandParametersBuilder {
    this.parameters.push({ required: true, names: [name] });
    return this;
  }

  withOptional(name: string): CommandParametersBuilder {
    this.parameters.push({ required: false, names: [name] });
    return this;
  }

  withRequiredOr(...names: string[]): CommandParametersBuilder {
    this.parameters.push({ required: true, names });
    return this;
  }

  withOptionalOr(...names: string[]): CommandParametersBuilder {
    this.parameters.push({ required: false, names });
    return this;
  }

  withDelimeter(delimeter: string): CommandParametersBuilder {
    this.delimeter = delimeter;
    return this;
  }

  build(): CommandParameters {
    return {
      parameters: this.parameters,
      delimeter: this.delimeter,
    };
  }

  buildEmpty(): CommandParameters {
    return {
      parameters: [],
      delimeter: null,
    };
  }
}

File: service/src/main/kotlin/xyz/lebkuchenfm/domain/commands/CommandProcessor.kt
package xyz.lebkuchenfm.domain.commands

import io.github.oshai.kotlinlogging.KLogger
import xyz.lebkuchenfm.domain.commands.model.Command
import xyz.lebkuchenfm.domain.commands.model.CommandProcessingResult

abstract class CommandProcessor(
    val key: String,
    val shortKey: String?,
    val helpMessage: String,
    val exampleUsages: List<String>,
    val parameters: CommandParameters,
) {
    abstract suspend fun execute(command: Command): CommandProcessingResult

    fun error(markdown: String, logger: KLogger, message: String? = null): CommandProcessingResult {
        logger.error { message ?: markdown }
        return CommandProcessingResult.fromMarkdown(markdown)
    }
}

class CommandParameters(
    val parameters: List<CommandParameter>,
    val delimiter: String = " ",
) {
    abstract class CommandParameter(
        val required: Boolean,
        val names: List<String>,
    )

    class RequiredCommandParameter(name: String) : CommandParameter(required = true, names = listOf(name))

    class OptionalCommandParameter(name: String) : CommandParameter(required = false, names = listOf(name))

    class RequiredMultiValue(vararg names: String) : CommandParameter(required = true, names.toList())

    class OptionalMultiValue(vararg names: String) : CommandParameter(required = false, names.toList())
}
